src/value/owned.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Provides owned implementations of [`SymbolToken`], [`Element`] and its dependents.
+//!
+//! This API is simpler to manage with respect to borrowing lifetimes, but requires full
+//! ownership of data to do so.
+
+use super::{AnyInt, Element, ImportSource, Sequence, Struct, SymbolToken};
+use crate::types::SymbolId;
+use crate::IonType;
+
+/// An owned implementation of  [`ImportSource`].
+#[derive(Debug, Clone)]
+pub struct OwnedImportSource {
+    table: String,
+    sid: SymbolId,
+}
+
+impl OwnedImportSource {
+    pub fn new<T: Into<String>>(table: T, sid: SymbolId) -> Self {
+        Self {
+            table: table.into(),
+            sid,
+        }
+    }
+}
+
+impl ImportSource for OwnedImportSource {
+    fn table(&self) -> &str {
+        &self.table
+    }
+
+    fn sid(&self) -> usize {
+        self.sid
+    }
+}
+
+/// An owned implementation of [`SymbolToken`].
+#[derive(Debug, Clone)]
+pub struct OwnedSymbolToken {
+    text: Option<String>,
+    local_sid: Option<SymbolId>,
+    source: Option<OwnedImportSource>,
+}
+
+impl OwnedSymbolToken {
+    pub fn new(
+        text: Option<String>,
+        local_sid: Option<SymbolId>,
+        source: Option<OwnedImportSource>,
+    ) -> Self {
+        Self {
+            text,
+            local_sid,
+            source,
+        }
+    }
+}
+
+impl<T: Into<String>> From<T> for OwnedSymbolToken {
+    /// Constructs an owned token that has only text.
+    fn from(text: T) -> Self {
+        Self::new(Some(text.into()), None, None)
+    }
+}
```

Comment:
We should provide a `From<SymbolId>` implementation too.

This raises an interesting question: when do we expect a `SymbolToken`'s text to be resolved? The `text()` method doesn't take a `SymbolTable`, so are we assuming it was resolved eagerly at `SymbolToken` creation time?

I can imagine a `LazySymbolToken` implementation that holds an `Rc<dyn SymbolTable>` pointer that's only read if `text()` is called.

Reply:
I don't really think the from `SymbolId` without either text or source is particularly useful though, because if that is true, you technically have a local symbol that has undefined text and that is equivalent to `$0`.

Reply:
This use case is roughly equivalent to a lazily resolved symbol token. I should be able to create a `SymbolToken` from `3usize` (i.e. a `SymbolId`) and give it to a writer without first providing or resolving the text. That doesn't mean there _isn't_ text for that symbol, but I haven't gone to the effort to look it up or otherwise provide it. (One could argue that we shouldn't facilitate this by providing an implementation of `From<SymbolId>` as a distinct concern. My point is that having a "raw" `SymbolToken` whose text is as of yet unknown is useful.)